pool/src/rate.rs

```diff
@@ -58,9 +61,80 @@ pub fn calc_interest_rate(
         return Some(max_rate);
     }
 
-    let base_rate_fixed = FixedI128::from_percentage(base_rate)?;
+    let initial_rate = FixedI128::from_percentage(ir_params.initial_rate)?;
 
-    let ir = base_rate_fixed.checked_div(denom)?;
+    let ir = initial_rate.checked_div(denom)?;
 
     Some(FixedI128::min(ir, max_rate))
 }
+
+/// Calculate accrued rate coefficient AR(t) = AR(t-1)*(1 + r(t-1)*elapsed_time)
+/// where:
+///     AR(t-1) - prev value of accrued rate
+///     r(t-1) - prev value of interest rate
+///     elapsed_time - elapsed time from last accrued rate update
+pub fn calc_accrued_rate_coeff(
+    prev_ar: FixedI128,
+    ir: FixedI128,
+    elapsed_time: u64,
+) -> Option<FixedI128> {
+    let delta_time = FixedI128::from_rational(elapsed_time, common::ONE_YEAR)?;
+    prev_ar.checked_mul(FixedI128::ONE.checked_add(ir.checked_mul(delta_time)?)?)
+}
+
+/// Calculates collateral and debt accrued coefficients and updates reserve data
+pub fn update_accrued_rates(
```

Comment:
Could you de-couple the storage logic from math?
I would suggest the following changes:
 - rename this method to `calc_accrued_rates` (to be consistent with `calc_interest_rate`) to return calculated ARs,
 - move `write_reserve(env, asset, &reserve_data);` to `get_actual_reserve_data`.

After that, `rate.rs` will contain only math.